stablehlo/tests/stablehlo_refine_shapes.mlir

```diff
@@ -0,0 +1,521 @@
+  %2 = stablehlo.constant dense<[1]> : tensor<1xi64>
+  %3 = stablehlo.real_dynamic_slice %arg0, %arg1, %1, %2
+           : (tensor<4xf32>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<*xf32>
+  func.return %3 : tensor<*xf32>
+}
+
+// -----
+
+// CHECK-LABEL: @refine_real_dynamic_slice_using_slice
+func.func @refine_real_dynamic_slice_using_slice(%arg0: tensor<4xf32>) -> tensor<*xf32> {
+  // CHECK: stablehlo.real_dynamic_slice{{.*}} -> tensor<1xf32>
+  %0 = stablehlo.constant dense<[0]> : tensor<1xi64>
+  %1 = stablehlo.constant dense<[1]> : tensor<1xi64>
+  %2 = stablehlo.constant dense<[1]> : tensor<1xi64>
+  %3 = stablehlo.real_dynamic_slice %arg0, %0, %1, %2
+           : (tensor<4xf32>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<*xf32>
+  func.return %3 : tensor<*xf32>
+}
+
+// -----
+
+// CHECK-LABEL: @refine_rng
+func.func @refine_rng(%arg0: tensor<f32>, %arg1: tensor<f32>) -> tensor<*xf32> {
+  // CHECK: stablehlo.rng{{.*}} -> tensor<4xf32>
+  %0 = stablehlo.constant dense<[4]> : tensor<1xi64>
+  %1 = stablehlo.rng %arg0, %arg1, %0, distribution = NORMAL : (tensor<f32>, tensor<f32>, tensor<1xi64>) -> tensor<*xf32>
+  func.return %1 : tensor<*xf32>
+}
+
+// -----
+
+// TODO(#1037): Switch to *x!quant.uniform<i8:f32, 1.000000e+00> once fixed.
+// CHECK-LABEL: @refine_uniform_quantize
+func.func @refine_uniform_quantize(%arg0 : tensor<4xf32>) -> tensor<?x!quant.uniform<i8:f32, 1.0>> {
+  // CHECK: stablehlo.uniform_quantize{{.*}} -> tensor<4x!quant.uniform<i8:f32, 1.000000e+00>>
+  %0 = stablehlo.uniform_quantize %arg0 : (tensor<4xf32>) -> tensor<?x!quant.uniform<i8:f32, 1.0>>
+  func.return %0 : tensor<?x!quant.uniform<i8:f32, 1.0>>
+}
+
+// -----
+
+// CHECK-LABEL: @refine_while
+func.func @refine_while(%arg0: tensor<4xf32>) -> tensor<*xf32> {
+  // TODO(#871): Also check cond and body when fixed.
+  // CHECK: stablehlo.while{{.*}} : tensor<4xf32>
+  // CHECK: stablehlo.abs{{.*}} : tensor<4xf32>
+  %0 = "stablehlo.while"(%arg0) ({
+  ^bb0(%arg1: tensor<?xf32>):
+    %1 = stablehlo.constant dense<true> : tensor<i1>
+    stablehlo.return %1 : tensor<i1>
+  },  {
+  ^bb0(%arg1: tensor<?xf32>):
+    stablehlo.return %arg1 : tensor<?xf32>
+  }) : (tensor<4xf32>) -> tensor<*xf32>
+  %1 = stablehlo.abs %0 : tensor<*xf32>
+  func.return %1 : tensor<*xf32>
+}
+
+// -----
+
+// TODO: Implement support for these ops.
+// * custom_call (#851).
+// * dynamic_conv (#867).
+
+// -----
+
+// CHECK-LABEL: func @update_function_type
+// CHECK-SAME: (%arg0: tensor<4xf32>) -> tensor<4xf32>
+func.func @update_function_type(%arg0: tensor<4xf32>) -> tensor<*xf32> {
+  // CHECK-NOT: tensor.cast
+  %0 = tensor.cast %arg0 : tensor<4xf32> to tensor<*xf32>
+  return %0 : tensor<*xf32>
+}
+
+// -----
+
+// CHECK-LABEL: func @update_region_type
+func.func @update_region_type(%arg0: tensor<i32>, %arg1: tensor<4xf32>) -> tensor<*xf32> {
+  // CHECK: "stablehlo.case"
+  // CHECK: -> tensor<4xf32>
+  // CHECK: stablehlo.abs{{.*}} : tensor<4xf32>
```

Comment:
I put this question in the "review cell" in last review and no "resolve" button related to that. Let me copy the question here again:

After looking at the tests, I'm wondering, will this PR breaks a lot of MHLO test? I mean we propagate the refined types in the later part of program (`abs` here) so the CHECKs will start to fail after the line of refined types because the output become different than what it was (tensor<*xf32> VS tensor<4xf32>? Does it need a switch to turn on/off these new passes for tests purpose?

Reply:
Given that this PR is limited to a single StableHLO pass (changes to `dialect/` are NFC along the lines of #1028), I don't expect that it will affect MHLO at all. I think we should be good.